### PR TITLE
:art: Refactor `ModelABC` to Help Use Default Torch Models

### LINKED ISSUE #566: Models inconsistency
- TIA Toolbox version: 1.3.3
- Python version: 3.10.8

### Description

Tiatoolbox has several pre-trained models helpful for data processing. However, models differ in how they handle input and output, making them confusing to use (especially when customizing):

- [ ] Activation functions apply at different steps for different models. Sometimes in `forward` method (e.g. [`CNNModel`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/2e8a0b6a5563346a8c644c1adb6ce90806efff3d/tiatoolbox/models/architecture/vanilla.py#L113)), sometimes `forward` returns a raw layer output and the transformation applies in `infer_batch` (e.g. [`UNetModel`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/1f0c8ead143cc65e325e924fd3c715017d3f4536/tiatoolbox/models/architecture/unet.py#L415)).
- [ ] Moreover, activation functions are hardcoded. To customize, you can't simply change an attribute; you must overwrite the whole method (different for each model).
- [ ] Data normalization distributes across methods: `HoVerNet` uses it in [`forward`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/b2556dd30694ff8840c2ebee7efffa337cdeaeb4/tiatoolbox/models/architecture/hovernet.py#L398), `UNetModel` in [`_transform`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/1f0c8ead143cc65e325e924fd3c715017d3f4536/tiatoolbox/models/architecture/unet.py#L338), `MicroNet` in [`preproc`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/d42f3238bdd4dea4f29483aa66140fa38c4fa5fe/tiatoolbox/models/architecture/micronet.py#L574), and vanilla models rely on the user to do so.
- [ ] Data preprocessing also lacks consistency. Even though it should happen in `preproc_func`/`_preproc` functions, `UNetModel` uses its own [`_transform`](https://github.com/TissueImageAnalytics/tiatoolbox/blob/1f0c8ead143cc65e325e924fd3c715017d3f4536/tiatoolbox/models/architecture/unet.py#L324), unrelated to the standard methods. Yet, its behavior could implement in `_preproc`.

### What to do
Refactoring the code will significantly improve readability:
- [ ] Decompose the pipeline into small granular methods in `ModelABC`: one method for normalization, activation function as an attribute, etc.
- [ ] Explain `ModelABC` methods in their documentation: does `infer_batch rely` on `postproc_func`? Can `infer_batch` be used for training? How?
- [ ] Reorganize custom model methods to match the new `ModelABC` structure.
- [ ] Add a new page to the documentation explaining the Tiatoolbox models pipeline: how is it related to the PyTorch pipeline? How to evaluate a model? How to train a model?

### PR DESCRIPTION
**Pull request by shaneahmed**
- Refactor `ModelABC` to Help Use Default Torch Models
- Remove `infer_batch` from `ModelABC`

### CODE REVIEW

File: tiatoolbox/models/architecture/vanilla.py [hunk 1]
```diff
@@ -78,6 +78,40 @@ def _get_architecture(
...
+    )  # to NCHW
+    img_patches_device = img_patches_device.permute(0, 3, 1, 2).contiguous()
+
+    # Inference mode
+    model.eval()
+    # Do not compute the gradient (not training)
+    with torch.inference_mode():
+        output = model(img_patches_device)
+    # Output should be a single tensor or scalar
+    return {"probabilities": output.cpu().numpy()}
```
Comment (GeorgeBatch) [inline tiatoolbox/models/architecture/vanilla.py:112]: In the current `develop` branch, neither `CNNModel`, nor `CNNBackbone` returned dictionaries as output of their `infer_batch()` methods. Also, `CNNModel` currently returns an array, while `CNNBackbone` returns a list with the array. It might be fine, just wanted to highlight this.

`CNNModel`
https://github.com/TissueImageAnalytics/tiatoolbox/blob/76f02f8561313dae786ebc15ab3196bde836a441/tiatoolbox/models/architecture/vanilla.py#L173

`CNNBackbone`
https://github.com/TissueImageAnalytics/tiatoolbox/blob/76f02f8561313dae786ebc15ab3196bde836a441/tiatoolbox/models/architecture/vanilla.py#L270

Comment (shaneahmed) [reply tiatoolbox/models/architecture/vanilla.py:113]: Thanks. We are aware of this. Our preference is to use torch nn models but to generalise for multi modal output we may need dictionaries. This PR is to check if we can move to generic torch models or we will need a sub class.

Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:113]: Makes sense.
